R/utils-get_code_dependency.R

```diff
@@ -0,0 +1,498 @@
+code_graph <- function(calls_pd) {
+
+  cooccurence <- extract_occurence(calls_pd)
+
+  side_effects <- extract_side_effects(calls_pd)
+
+  graph <- mapply(c, side_effects, cooccurence, SIMPLIFY = FALSE)
+
+  remove_graph_duplicates(graph)
+
+}
+
+#' @keywords internal
+#' @noRd
+extract_occurence <- function(calls_pd) {
+
+  used_in_function <- function(call, rows) {
+
+    # If an object is a function parameter,
+    # then in calls_pd there is a `SYMBOL_FORMALS` entry for that object.
+    objects <- sapply(rows, function(row) call$text[row])
+
+    vapply(objects, function(object) {
+      if (any(call[call$token == "SYMBOL_FORMALS", "text"] == object) && any(call$token == "FUNCTION")) {
+        object_sf_ids <- call[call$text == object & call$token == "SYMBOL", "id"]
+        function_start_id <- call[call$token == "FUNCTION", "id"]
+        all(object_sf_ids > function_start_id)
+      } else {
+        FALSE
+      }
+    },
+      logical(1)
+    )
+  }
+
+  extract_functions <- function(calls_pd) {
+    # Detect functions created within code.
+    pd <- do.call(rbind, calls_pd)
+    symbols <- pd[pd$token %in% c("SYMBOL", "SYMBOL_FUNCTION_CALL"), c('token', 'text')]
+    symbols_table <- table(unique(symbols)$text)
+    names(symbols_table[symbols_table == 2])
+  }
+
+  functions <- extract_functions(calls_pd)
+
+  lapply(
+    calls_pd,
+    function(x) {
+      sym_cond <- which(x$token == "SYMBOL" | (x$token == "SYMBOL_FUNCTION_CALL" & x$text %in% functions))
+      sym_cond <- sym_cond[!used_in_function(x, sym_cond)]
+
+      # Watch out for SYMBOLS after $ and @. For x$a x@a: x is object, a is not.
+      # For x$a, a's ID is $'s ID-2 so we need to remove all IDs that have ID = $ID - 2.
+      dollar_ids <- x[x$token %in% c("'$'", "'@'"), "id"]
+      if (length(dollar_ids)) {
+        object_ids <- x[sym_cond, "id"]
+        after_dollar <- object_ids[(object_ids - 2) %in% dollar_ids]
+        sym_cond <- setdiff(sym_cond, which(x$id %in% after_dollar))
+      }
+
+      # If there was an assignment operation detect direction of it.
+      if (length(sym_cond) >= 2) {
+        ass_cond <- grep("ASSIGN", x$token)
+        if (length(ass_cond)) { # NOTE(1)
+          sym_cond <- sym_cond[sym_cond > ass_cond] # NOTE(2)
+          text <- x[sort(c(sym_cond, ass_cond)), "text"]
+
+          ans <- if (text[1] == "->") { # NOTE(3)
+            rev(text[-1])
+          } else {
+            text[-1]
+          }
+          c(ans[1], ":", unique(ans[-1]))
+        } else {
+          x[sym_cond, "text"]
+        }
+      } else {
+        x[sym_cond, "text"]
+      }
+
+      ### NOTE(3): What if there are 2+ assignments, e.g. a <- b -> c or e.g. a <- b <- c.
+      ### NOTE(2): For cases like 'eval(expression(b <- b + 2))' removes 'eval(expression('.
+      ### NOTE(1): Cases like 'data(iris)' that do not have an assignment operator.
+    }
+  )
+}
+
+#' @keywords internal
+#' @noRd
+extract_side_effects <- function(calls_pd) {
```

Comment:
please expand docs of this function - please explain side effects term

Reply:
03710f3e41b27f47e0d2fcbf0b5650df5297ef70